### PR TITLE
Use a class for help command and do some fixes to tests

### PR DESCRIPTION
**Pull request by voxsim**
This is my first tiny step towards a Base Command class to clash the number of ifs in src/cli/index.js. The way is quite long :D

**Summary**
For now I fixed some tests and write the help command as es6 class.

**Test plan**
I fixed some tests before touching some code; please see the commits.
Added tests for cache command.

@bestander I hope you like what I do and I have some questions for you:
1. Do you think is useful add an acceptance/feature test per command? It maybe be overkill but the code is too much entagled and if you agree we can remove in a second moment after the story ends if they are useless.
2. I think I will open an issue and maybe I will do a checklist because it is quite long and maybe it is useful some feedbacks/help, what do you think?

Right I am working to have all classes with three static functions (run, hasWrapper, setFlags), and when I will be ready everything will be switched to instance function. I will open more pull request to be careful ;)

### CODE REVIEW

File: src/util/misc.js [hunk 1]
```diff
@@ -62,3 +62,13 @@ export function camelCase(str: string): ?string {
     return _camelCase(str);
   }
 }
+
+export function findIndex(args: Array<string>, predicate: string): number {
```
Comment (bestander) [inline src/util/misc.js]: Why don't you use `args.findIndex()`?

**Comment by bestander** [comment]
@voxsim, thanks for bringing this up.
Let's open an issue to track gradual improvement.
This may require a few PRs, we need to split it into atomic steps.

File: src/util/misc.js (hunk 1 above)
Comment (voxsim) [inline src/util/misc.js]: Ops you are right ^^' I really don't remember why did this

**Comment by bestander** [comment]
Ping me when it is ready for review

**Comment by voxsim** [comment]
@bestander I think the code is ready for review; the failure of travis does not seem related to my last changes O.o

**Comment by voxsim** [comment]
I added other commits because I had time and maybe you can give me some feedbacks on how I arrange the cache command class (before it was using buildSubCommand, maybe it is worse than before for you).

File: __tests__/index.js [hunk 2]
```diff
@@ -204,3 +204,9 @@ if (process.platform !== 'win32') {
     expect(stdout[stdout.length - 2]).toEqual('A message from custom script with args --help');
   });
 }
+
+test.concurrent('should run bin command', async () => {
+  const stdout = await execCommand('bin', [], '', false);
+  expect(stdout[0]).toEqual(`${fixturesLoc}/node_modules/.bin`);
```
Comment (bestander) [inline __tests__/index.js]: this test fails on windows CI (appveyor)

Comment (voxsim) [inline __tests__/index.js]: Oh what a newbie error ^^' i'll fix asap!

**Comment by voxsim** [comment]
@bestander AppVeyor breaks for both builds for:

`Timeout - Async callback was not invoked within timeout specified by jasmine.DEFAULT_TIMEOUT_INTERVAL`

That does not seem related to my changes (I haven't changed anything related to the installation command). Could you please check? I'll investigate further in the next days.

I added some tests for the cache command, now every change is tested properly!

**Comment by voxsim** [comment]
Maybe this pull request https://github.com/yarnpkg/yarn/pull/2887 is better. I will close this after discussing which is better.

**Comment by bestander** [comment]
I'll assign @arcanis as reviewer of this and the alternative PR.
Good work, @voxsim, sorry for the delays, we are just trying to balance between fixing bugs, adding new features and improving code maintainability.
Sometimes the latter gets a lower priority.

**Comment by bestander** [comment]
Overall I like the BaseCommand/SubCommand idea but I let @arcanis drive this with you

**Comment by voxsim** [comment]
after some parallel work i want to address this in another way ;) tomorrow or sunday I will write something here or write another pull request.

Btw do you need some help about addressing issues or bugs? I can help, sorry for the last month but I was busy for personal reasons, but I should come back :p

**Comment by bestander** [comment]
Yes yes yes! :)

We definitely need help with triagibg bugs and reviewing PRs.
Just go ahead and review PRs and ping me for merges or access to labelling.

We usually give permissions to active members after they've sent a dozen
PRs and shown care for the project.

**Comment by voxsim** [comment]
ok ;)

**Comment by voxsim** [comment]
Right now I am closing this pull request in favor of https://github.com/yarnpkg/yarn/pull/3105. Maybe in the future we will extract a base class, we will see :)

**Pull request closed without merging**